research/2019-12-16_searching_with_the_calm_api.md

```diff
@@ -0,0 +1,241 @@
+              f"{base_api_url}/CalmAPI/ContentService.asmx",
+              headers={
+                  "SOAPAction": "http://ds.co.uk/cs/webservices/SummaryHeader",
+                  "Content-Type": "text/xml; charset=utf-8"
+              },
+              auth=auth,
+              data=f"""
+              <?xml version="1.0" encoding="utf-8"?>
+              <soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
+                <soap:Body>
+                  <SummaryHeader xmlns="http://ds.co.uk/cs/webservices/">
+                    <dbname>{database}</dbname>
+                    <HitLstPos>{hit_lst_pos}</HitLstPos>
+                  </SummaryHeader>
+                </soap:Body>
+              </soap:Envelope>
+              """.strip()
+          )
+
+          # The response XML is of the form:
+          #
+          #     <?xml version="1.0" encoding="utf-8"?>
+          #     <soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" ...>
+          #       <soap:Body>
+          #         <SummaryHeaderResponse xmlns="http://ds.co.uk/cs/webservices/">
+          #           <SummaryHeaderResult>string</SummaryHeaderResult>
+          #         </SummaryHeaderResponse>
+          #       </soap:Body>
+          #     </soap:Envelope>
+          #
+          result_root = ET.fromstring(summary_header_resp.content)
+
+          result_string = result_root.find(
+              "./"
+              "{http://schemas.xmlsoap.org/soap/envelope/}Body/"
+              "{http://ds.co.uk/cs/webservices/}SummaryHeaderResponse/"
+              "{http://ds.co.uk/cs/webservices/}SummaryHeaderResult"
+          ).text
+
+          # The result is of the form:
+          #
+          #     <?xml version="1.0" encoding="ISO-8859-1"?><SummaryList><Summary>
+          #       <RecordType>Component</RecordType>
+          #       <IDENTITY></IDENTITY>
+          #       ...
+          #     </Summary></SummaryList>
+          #
+          summary_root = ET.fromstring(result_string).find(".//Summary")
+          yield summary_root
+      ```
+    </details>
+
+The Search request is attached to a persistent "session" -- somehow the API knows you want the results from a given Search request (cookies, maybe).
+I was using a persistent `requests.Session()` in Python and that seemed to work fine.
+
+Two possible approaches:
+
+*   **I want to get everything.**
+
+    Make a Search request with database `Catalog` and search term `*`.
+    That tells you how many results there are, then iterate from `0 to (hit count – 1)` (the results are 0-indexed), and call SummaryHeader for each.
+
+*   **I want to get recently updated stuff.**
+
+    You don't want to make a SummaryHeader request for each record -- that's slow and unnecessary.
+
+    Make a Search request with database `Catalog` and search term `*`.
```

Comment:
* as a wildcard sometimes gives slightly odd results in the Calm client, so it might be worthwhile comparing the results of this search in the API with a similar search in the client, to check they are as expected. * in the client only matches 'text-type' fields and only where the field is not empty, and there's a separate keyboard shortcut (which must map to something programatically but I don't know what) for 'find all records in database'. As there are no field(s) specified here, maybe this isn't an issue though... but then what is the purpose of the <elementSet>DC</elementSet> below?

Reply:
The `<elementSet>DC</elementSet>` is something the CALM API docs suggest; I haven't experimented to see what happens if you omit that field.